packages/node/src/retryHandler.ts

```diff
@@ -0,0 +1,174 @@
+          }
+        });
+      }
+    } finally {
+      return response;
+    }
+  }
+
+  private _setupTransport(): Transport {
+    let transportOptions: TransportOptions;
+    transportOptions = {
+      serverUrl: this._options.serverUrl || AMPLITUDE_SERVER_URL,
+      headers: {
+        'Content-Type': 'application/json',
+      },
+    };
+    return new HTTPTransport(transportOptions);
+  }
+
+  private _shouldAttemptRetry(): boolean {
+    if (typeof this._options.maxRetries === 'number' && this._options.maxRetries <= 0) {
+      return false;
+    }
+
+    const bufferLimit = this._options.maxCachedEvents ?? 100;
+
+    return this._eventsInRetry < bufferLimit;
+  }
+
+  // Sends events with ids currently in active retry buffers straight
+  // to the retry buffer they should be in
+  private _pruneEvents(events: Array<Event>): Array<Event> {
+    const prunedEvents: Array<Event> = [];
+    events.forEach(event => {
+      const id = this._getId(event);
+      if (id) {
+        const retryBuffer = this._idToBuffer.get(id);
+        if (retryBuffer?.length) {
+          retryBuffer.push(event);
+          this._eventsInRetry++;
+        } else {
+          prunedEvents.push(event);
+        }
+      }
+    });
+
+    return prunedEvents;
+  }
+
+  private _getId(event: Event): string {
+    // events should either have user or device id
+    if (typeof event.user_id === 'string') {
+      return event.user_id;
+    } else if (typeof event.device_id === 'string') {
+      return event.device_id;
+    } else {
+      return '';
+    }
+  }
+
+  private _getPayload(events: Array<Event>): Payload {
+    return {
+      api_key: this._apiKey,
+      events,
+    };
+  }
+
+  private _cleanUpBuffer(id: string): void {
+    const eventsToRetry = this._idToBuffer.get(id);
+    if (!eventsToRetry) {
+      return;
+    } else if (!eventsToRetry.length) {
+      this._idToBuffer.delete(id);
+      return;
+    }
+  }
+
+  private async _retryEvents(id: string): Promise<void> {
+    const eventsToRetry = this._idToBuffer.get(id);
+    if (!eventsToRetry?.length) {
+      this._cleanUpBuffer(id);
+      return;
+    }
+
+    const initialEventCount = eventsToRetry.length;
+
+    let numRetries = 0;
+    const maxRetries = this._options.maxRetries ?? 0;
+
+    while (numRetries < maxRetries) {
+      // If new events came in in the meantime, collect them as well
+      const arrayLength = eventsToRetry.length;
+      if (arrayLength === 0) {
+        return;
+      }
+
+      try {
+        const response = await this._transport.sendPayload(this._getPayload(eventsToRetry));
+        if (response.status === Status.Success) {
+          // Clean up the events
+          eventsToRetry.splice(0, arrayLength);
+          this._eventsInRetry -= arrayLength;
+          // Successfully sent the events, stop trying
+          break;
+        } else {
+          throw new Error(response.status);
+        }
+      } catch {
```

Comment:
I think the specific case you mention should be a utility that can be caught just as the nodeClient's `logEvent` so I think it might make sense to have a `validateEvent()` step there as well. we can add that check to the events as they're about to be sent as well to make sure those events are not included in the payload.

Reply:
I think a validate step in the `logEvent` call that throws immediately could be helpful but I think the retry handling here should be considered separately.

We can't guarantee that users will update the SDK even if we try to keep it in sync with all the validation that the event servers do. So we should expect the possibility that the event server will return a 400 for a reason the SDK couldn't validate. We should handle those cases differently from 429s or 5xxs or local network errors.